### ISSUE TITLE
No animation when adding/removing series in highstock

### ISSUE THREAD

**Issue by arnaudruffin** [highcharts/highcharts#4406]
Hello

I try to dynamically add/remove series in a highstock chart. I have tried various options, but I can't get the added series to animate. 

The only animation showing are:
- the first serie plotting during chart creation
- the zoom/scale animation when adding data

Here is what it looks like:
http://jsfiddle.net/mtasxa0o/

i even tried to force animation when adding a serie `mchart.addSeries({},true,true);` but with no luck.

**This only applies to highstock.** If I call the same code with highchart it works:
http://jsfiddle.net/rjj4hL59/

Am I missing something?

**Comment by pawelfus** [comment]
Workaround:
Set animation.duration for series **and** for animation param in `addSeries`: http://jsfiddle.net/mtasxa0o/1/

Looks like duration needs to be different that any used before (?), see: http://jsfiddle.net/mtasxa0o/2/

**Comment by arnaudruffin** [comment]
Thanks for the workaround!
Really curious behaviour though :S.

I've also noted that this animations settings must be set directly in the serie:
http://jsfiddle.net/mtasxa0o/3/

Passing a parameter to addSerie method won't do anything:
http://jsfiddle.net/mtasxa0o/4/

**Comment by pawelfus** [comment]
Good point! Animation for series is enough. I guess this may be connected to the fact, that initial animation is a bit different than all others.

**Comment by Narretz** [comment]
According to the docs: http://api.highcharts.com/highcharts/Chart.addSeries, giving `true` or an object with a duration prop, should definitely animate addSeries, even if no or a different animation is set in the config. So either the docs are wrong or this is broken.

**Comment by rwang1031** [comment]
Hello guys:
        After played around a little with the latest highstock.src.js
        I found that if you make just a little change the a priavte function called  'setClip', the problem will be fixed.  
         Solution: Change the flag in the if condition from 'clipRect' to 'markerClipRect'; (line number:23860)
            if (!clipRect) {  // change this line to  if (!markerClipRect) {

                    // When animation is set, prepare the initial positions
                    if (animation) {
                        clipBox.width = 0;

                        chart[sharedClipKey + 'm'] = markerClipRect = renderer.clipRect(-99, // include the width of the first marker
                            inverted ? -chart.plotLeft : -chart.plotTop,
                            99,
----------------------------------------------------------------------------------------------
The reason why the newly added series doesn't animate is that: 
after the first added series is animated, it triggers it's afterAnimation event which calls the setClip function again:
------------------------------------------------
   afterAnimate: function() {
                this.setClip();
                fireEvent(this, 'afterAnimate');
                this.finishedAnimating = true;
            },
---------------------------------------------
And then the setClip() uses clipRect object as a flag to differenciate it from the first call.
But when you add the second series, if it has the same duration as the first one, it will have the same  sharedClipKey as the first one:
------------------------------------------------

this.sharedClipKey || [
                        '_sharedClip',
                        animation && animation.duration,
                        animation && animation.easing,
                        clipBox.height,
                        options.xAxis,
                        options.yAxis
                    ].join(',')
------------------------------------------------------------
So that the clipRect object won't be null,  so the setClip() doesn't know if it's trigger by the addseries event or afterAnimate event.

But the markerClipRect object doesn't have this problem, because it's being destroyed every time during the afterAnimation call. Not like the clipRect which doesn't get destroyed unless clipBox of the chart is destroyed.
-----------------------------------------

                    if (clipRect.count.length === 0 && sharedClipKey && chart[sharedClipKey]) {
                        if (!seriesClipBox) {
                            chart[sharedClipKey] = chart[sharedClipKey].destroy();
                        }
                        if (chart[sharedClipKey + 'm']) {
                            chart[sharedClipKey + 'm'] = chart[sharedClipKey + 'm'].destroy();
                        }
                    }
-------------------------------------------------------
I've played around for a while after this change to the source code, haven't found any side effect. But it's still hard to tell.

I hope it's helpful, good luck!

**Comment by robhunt3r** [comment]
Any news on this? I would love to see my easing when updating my series data/removing it and adding it again.

When updating, the effect is ugly (it just fills the left side poorly) and when removing/adding it animates by defalut, not with my easings :C

**Comment by pawelfus** [comment]
No news @arnaudruffin - let me reopen this ticket. Did you try my workarounds posted above? Note that initial animation should be set in `series.animation`, like this: http://jsfiddle.net/BlackLabel/hod3q57y/

**Comment by lukasschueler** [comment]
Hey guys, 
I'm confronted with the same problem right now. Anything new on this?
Additionally, for me, in all of your fiddles only the first plot animation works. The other appear instantaneous.

**Comment by pawelfus** [comment]
That's quite interesting: bug was reported in 2015, but just tested some v6.0 - v8.0 releases and all of them were fine.

Anyway, last working version in v8 family is v8.0.2: http://jsfiddle.net/BlackLabel/o9vrd5zm/1/

Bisected to commit 5b06bed1a7

Issue related to #13751 

**Workaround:**
Use the same as in #13751 and enable initial animation.
Demo: http://jsfiddle.net/BlackLabel/o9vrd5zm/3/

**Comment by mateuszkornecki** [comment]
The same thing is happening in the current highcharts without the stock module imported.

**Live demo:**
https://jsfiddle.net/BlackLabel/26a4sjrm/

**Comment by KacperMadej** [comment]
For Highcharts works fine for the next release candidate (at the time of writing the current version is 8.2.2): https://jsfiddle.net/BlackLabel/fds3v086/

The problem is still visible for Highcharts Stock: https://jsfiddle.net/BlackLabel/fds3v086/1/

**Comment by pawelfus** [comment]
Refactoring is needed first (https://github.com/highcharts/highcharts/pull/15718#issuecomment-876992475)
